fs: extends recursive readdir to allow a function to limit traversal

prettydiff opened the pull request:
This commit contains the necessary fixes to code and documentation per manual testing, but I still need to complete test units for test autoamtion.  This pull request does not claim to be complete until test automation is updated, but is enough to open discussion for a new feature.

fixes: https://github.com/nodejs/node/issues/49243

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

bricss commented:
IMO, it's gonna be semver major unless *depth* is set to `Infinity` ♾️ by default 🎚️

prettydiff commented:
@bricss That should not be an issue.  There is some complexity around the default value because I am trying to not interfere with the existing functionality of option `recursive`.  Otherwise, I would just replace option `recursive` entirely.

The values for `depth` are:
* less than 1 fully recursive, as in infinity
* 1 - no recursion
* greater than 1 the number of recursive steps to take

If there is no indication of recursion, such as when both options `recursive` and `depth` are absent, the default value of `depth` is 1 for no recursion.  Otherwise the default value is 0 so that depth does not interfere with the full recursion provided by option `recursive`.

aduh95 commented on lib/fs.js:
```diff
@@ -1498,7 +1506,13 @@ function readdir(path, options, callback) {
   }
 
-  if (options.recursive) {
+  const depth = (options !== null && options !== undefined && typeof options.depth === 'number') ?
+    MathFloor(options.depth) :
+    options.recursive ?
+      0 :
+      1;
```
```suggestion
  const depth = options?.depth ?? (options?.recursive ? Infinity : 1);
  if (depth !== null && depth !== Infinity) {
   validateInteger(depth, 'options.depth', 1);
 }
```

aduh95 commented on doc/api/fs.md:
```diff
@@ -1305,6 +1305,7 @@ changes:
   * `encoding` {string} **Default:** `'utf8'`
   * `withFileTypes` {boolean} **Default:** `false`
   * `recursive` {boolean} **Default:** `false`
+  * `depth` {number} **Default:** `1`
```
```suggestion
  * `depth` {integer} **Default:** `1`
```

aduh95 commented on doc/api/fs.md:
```diff
@@ -1318,6 +1319,12 @@ will be passed as {Buffer} objects.
 If `options.withFileTypes` is set to `true`, the resolved array will contain
 {fs.Dirent} objects.
 
+The `options.depth` determines the amount of recursion to apply. A value less
+than 1 applies full recursion, value of 1 applies no recursion, and greater
```
```suggestion
The `options.depth` determines the amount of recursion to apply. A value of
`Infinity` applies full recursion, value of 1 applies no recursion, and greater
```

aduh95 commented on doc/api/fs.md:
```diff
@@ -1318,6 +1319,12 @@ will be passed as {Buffer} objects.
+than 1 applies full recursion, value of 1 applies no recursion, and greater
+values determine the depth of descendant directies to traverse. If
+`options.recursive` is value `true` then `options.depth` defaults to a value
+of `0` and otherwise defaults to a value `1`.
```
```suggestion
of `Infinity` and otherwise defaults to a value `1`.
```

aduh95 commented on lib/fs.js:
```diff
@@ -1469,9 +1471,15 @@ function readdirSyncRecursive(basePath, options) {
+    if (
+      options.depth < 1 ||
+      pathsQueue[i] === basePath ||
+      pathsQueue[i].replace(base, '').split(sep).length < options.depth
```
This likely has a not so great performance, couldn't we keep track of how deep we are instead using a variable that we increment?

prettydiff replied:
The current implementation makes this confusing.  The async function fs.readdir currently uses the synchronous `readdirSyncRecursive` to apply recursion.  In my own personal library where I worked this out in the past its always asynchronous because I perform a different fs.readdir on each child directory, so I expect performance to differ there versus here.  In my own code though, this approach proved insignificantly slow because its always less slow than transitions through the event loop on each readdir call.

In order to keep track of depth you would have to keep track per each descendant directory.  That is why implemented depth checks in the first place, because such a tracking tree can become exceedingly large.

aduh95 commented:
/cc @nodejs/fs

bnoordhuis commented:
Instead of adding another special-purpose option, why not a `shouldContinue` or `shouldEnter` callback (actual name TBD) that's invoked whenever node is about to descend into a new directory? Mechanism > policy, always.

prettydiff commented:
@bnoordhuis I do not understand your guidance and request an example.

tniessen commented:
I believe what Ben means is that it is not a good pattern to keep adding options to support different use cases. For example, someone else might want to filter recursion by depth, path, directory names, whether something is a symbolic link, etc. Instead, a mechanism that allows users to define their own criteria would be desirable.

prettydiff commented:
To accomplish that I can eliminate the new `depth` option and change the `recursive` to support search criteria.  I was trying to avoid something like this because it means changing the API and becomes semver major.

I propose the `recursive` option changes from a boolean type to a key/value pair where the key determines the search type and the value determines what to search for.  Here are the keys I propose and more can be added into the future:

* count {integer} - halt when the result list length reaches the supplied value
* depth {integer} - how far to traverse
* fragment {string} - a file system search on string fragment
* regexp {RegExp} - a regular expression search
* negative {string} - a negative fragment search, as in items to disqualify from the result list

I have each of these working already in my personal code, except for count, and it executes very quickly.  I have to dive a bit further into how the recursive lists are actually generated in the Node library.

aduh95 commented:
IIUC Ben's suggestion is to accept a function that would return a boolean, I guess something that would look like

```js
readdir('./someDir', {
  recursive(dir, depth) {
    // If the function returns `true`, the recursion keeps going deeper.
    return depth < 3 && dir.name.startsWith('someString');
  },
})
```

Then users can implement filters as complicated or as simple as they want.

prettydiff commented:
Options `depth` converted to `recursiveFilter` which accepts a function.  The function expects 1 string argument and expects to return a boolean.

```javascript
fs.readdir("path", {
    recursive: true,
    recursiveFilter: function (path) {
        return path.indexOf("vi") > 0;
    }
})
```

This imposes different behavior.
1. Option recursiveFilter is ignored unless option `recursive` is true.
2. Applies to all paths irrespective of file system type, as in a name search, opposed to operating only upon directory types as in a navigational limit.

bricss commented:
It would be great to change `recursiveFilter` to pass [dir](https://nodejs.org/api/fs.html#class-fsdir) as the first and `path` as the second argument so there more options for filtration 🛃

LiviaMedeiros commented:
How about this: if `options.withFileTypes` is set to false, we pass (`path`, `depth`); and if it's true, we pass ([`dirent`](https://nodejs.org/api/fs.html#class-fsdirent), `depth`). This would be more aligned with returned value and probably synergized with common usecases.

To solve the depth limiting, IMHO there definitely should be an integer `depth` counter
```mjs
{ recursiveFilter: (path, depth) => depth < 3 }
// much cleaner than something like:
{ recursiveFilter: (path) => path.split(pathModule.sep).length < 4 }
```

Aside from that, wouldn't it make more sense to make the callback determine "should we descend into this directory" or "should we continue altogether" rather than "should we process&return this file or directory"?
With filter function, e.g. in `(_, depth) => depth < 3` scenario, it still will be reading all directories with `depth === 2` just to immediately filter out everything in them, which might be a lot of futile I/O operations.
Filtering regular files can be done with `resultArray.filter()` afterwards, there is not much benefit in doing that internally.

prettydiff commented:
Updated to pass in an object containing all the desired data and the code is cleaner.

```javascript
fs.readdir("path", {
    recursive: true,
    recursiveFilter: function (filterOutput) {
        return filterOutput.parentName.indexOf("vi") > 0;
    }
})
```

Where `filterOutput` returns this object:

```javascript
{
    depth: {number}, // steps from the supplied base path
    itemName: {string}, // name of the given file system artifact
    itemPath: {string}, // absolute path of the given file system artifact
    parentName: {string}, // name of the parent directory
    parentPath: {string} // absolute path of the parent directory
}
```

LiviaMedeiros commented on lib/internal/fs/promises.js:
```diff
@@ -792,48 +794,72 @@ async function readdirRecursive(originalPath, options) {
+  const lengthReg = (sep === '\\') ?
+    new RegExp('\\\\', 'g') :
+    new RegExp('/', 'g');
+  const baseDirLength = originalPath.match(lengthReg).length;
```
Wouldn't it be better to start with `let depth = 0;` and use `[path, readdir, depth + 1]` as queue item?

prettydiff replied:
In this approach the base path passed into the recursive function is depth 0.  It's immediate children are depth 1, and the immediate children of the immediate child directories are depth 2, and so forth.  It is counting the number of steps removed from the specified address.

LiviaMedeiros replied:
Sorry, I should've worded it better.
I meant something like:
```mjs
// base directory inits with depth = 0
const queue = [
  [
    originalPath,
    await binding.readdir(...),
    0, // initial depth
  ],
];
```
```mjs
// processing each element from queue
const { 0: path, 1: readdir, 2: depth } = ArrayPrototypePop(queue);
for (const ent of readdir) {
// ...
  if (stat === 1) {
    ArrayPrototypePush(queue, [
      direntPath,
      await binding.readdir(...),
      depth + 1, // subdirectories will have 1, subsubdirectories will have 2, etc.
    ]);
// ...
```
This would achieve the same thing with better performance and robustness, avoiding parsing path every time, counting separators, and memorizing `baseDirLength`.

LiviaMedeiros commented on lib/internal/fs/promises.js:
```diff
@@ -792,48 +794,72 @@ async function readdirRecursive(originalPath, options) {
-              kUsePromises,
-            ),
-          ]);
+      if (filter(pathModule.resolve(path))) {
+        for (const dirent of getDirents(path, readdir)) {
+          if (filter(dirent)) {
```
It seems that this code will call `filter()` twice for processed directories.

prettydiff replied:
The first time filters for directory traversal and the second time is for items in the given directory.

LiviaMedeiros replied:
Which means, every subdirectory will be passed to `filter()` as item of its parent, and then again as current directory.

IMHO, one filter here
```diff
-if (dirent.isDirectory()) {
+if (dirent.isDirectory() && filter(dirent, depth)) {
```
would be sufficient in case if we want to be able to prevent further `readdir` calls by returning falsey value, and returning regular files unconditionally.

LiviaMedeiros commented on lib/internal/fs/promises.js:
```diff
@@ -792,48 +794,72 @@ async function readdirRecursive(originalPath, options) {
+    const directory = absolutePath.slice(0, absolutePath.lastIndexOf(sep));
+    return absolutePath === originalPath ||
+      typeof options.recursiveFilter !== 'function' ||
+      (typeof options.recursiveFilter === 'function' && options.recursiveFilter({
```
In `fsPromises.readdir()` version, can `recursiveFilter` be async?

prettydiff replied:
I do not see the need.  There are no events or external actions here that would hang the thread.  Forcing it to become async would push it into another callstack and there is a minor performance penalty for doing so.

LiviaMedeiros replied:
Perhaps. I wondered if, for example, user would want to run `await fsPromises.stat(dir)` inside filter function.

LiviaMedeiros reviewed (commented):
IMHO it would be better to keep `readdir()` filter callback as simple as possible: call it only for directories, determine only if we should open them (but not if we want to keep them in returned array), and pass only `path`/`Dirent` and `depth`.

If we want to make a convenience method that will utilize proposed `fs.FilterOutput` or even more values such as `size`, `mtime`, `extname` etc. etc., I think it's better to introduce separate `fs.walk()` API that would also filter regular files and return `Iterable` or `ReadableStream` instead of plain array.

prettydiff commented:
@LiviaMedeiros That would make things a tiny bit more simple because directory traversal is isolate from child item population in the existing logic.  I can solve for that now by just removing code from the current approach, but I am going to spend some to think through this in case I am missing something or if somebody else has a different opinion.

tniessen commented on doc/api/fs.md:
````diff
@@ -6546,6 +6582,39 @@ added:
 
 The base path that this {fs.Dirent} object refers to.
 
+### Class: `fs.FilterOutput`
+
+A {fs.FilterOutput} object provides identifiers for file system artifacts. This
+object is passed into the function provided to option `recursiveFilter` of
+[`fsPromises.readdir()`][], [`fs.readdir()`][], and [`fs.readdirSync()`][] when
+option `recursive` is assigned `true`.
+
+```console
+FilterOutput {
+  depth: 2,
+  itemName: 'fileName.ts',
+  itemPath: '/users/userAccount/notes/fileName.ts',
+  parentName: 'notes',
+  parentPath: '/users/userAccount/notes'
+}
+```
````
Most of this seems pretty redundant. For consistency with `fs.cp()`, why not just pass the (potentially relative) path to the function? Determining the depth is simple enough if necessary, and so is determining the `parentName` etc.

prettydiff commented:
I am taking the advice from @LiviaMedeiros and simplifying this effort as much as possible.  There will be a depth option and it determines the number of descendant directories to traverse such that 0 eliminates recursion, 1 traverses the immediate child directories, and so forth.  I will look at proposing a `fs.walk` that does wonderfully advanced search things later as suggested.

Sorry I took days to respond.  I am living in a hotel and my ethernet port died.

prettydiff commented:
Effort complete, test units added, all build tasks green!

tniessen commented on doc/api/fs.md:
```diff
@@ -1318,6 +1319,11 @@ will be passed as {Buffer} objects.
 
+`options.depth` only applies if option `options.recursive` is set to `true` and
+determines how many descendant depths of file system artifacts to include. A
+value of 0 eliminates recursion while a value of 1 traverses immediately child
```
```suggestion
value of 0 eliminates recursion while a value of 1 traverses immediate child
```

tniessen commented on doc/api/fs.md:
```diff
@@ -3633,6 +3640,11 @@ the filenames returned will be passed as {Buffer} objects.
 
+`options.depth` only applies if option `options.recursive` is set to `true` and
+determines how many descendant depths of file system artifacts to include. A
+value of 0 eliminates recursion while a value of 1 traverses immediately child
```
```suggestion
value of 0 eliminates recursion while a value of 1 traverses immediate child
```

tniessen commented on doc/api/fs.md:
```diff
@@ -5695,6 +5708,11 @@ the filenames returned will be passed as {Buffer} objects.
 
+`options.depth` only applies if option `options.recursive` is set to `true` and
+determines how many descendant depths of file system artifacts to include. A
+value of 0 eliminates recursion while a value of 1 traverses immediately child
```
```suggestion
value of 0 eliminates recursion while a value of 1 traverses immediate child
```

tniessen commented on lib/fs.js:
```diff
@@ -1452,7 +1469,7 @@ function readdirSyncRecursive(basePath, options) {
         const dirent = getDirent(path, readdirResult[0][i], readdirResult[1][i]);
         ArrayPrototypePush(readdirResults, dirent);
-        if (dirent.isDirectory()) {
+        if (dirent.isDirectory() && depthCheck(readdirResult[0][i])) {
```
Isn't `readdirResult[0][1]` potentially a `Buffer`?

prettydiff replied:
According to the comment immediately preceding this logic readdirResult is an array of two arrays where the first is the name of file system artifacts and the second is the type.  Can a file system item have a buffer as a name?  If so I will add a `.toString()` to ensure coercion to a string unless there is a better way to handle this.  The value that is passed in will be converted to an absolute path.

tniessen replied:
> Can a file system item have a buffer as a name?

Yes, if you set the `encoding` option to `'buffer'`.

> If so I will add a `.toString()` to ensure coercion to a string

`toString()` will decode the `Buffer` as a sequence of UTF-8 code points. However, as far as I can tell, there is no guarantee that the name will actually be a UTF-8 string.

> unless there is a better way to handle this.

I am not aware of any way to handle this. The `node:fs` module used to be a nice low-level interface for well-defined syscalls that would let the operating system handle paths.

prettydiff replied:
In that case it sounds like there is not a good way to account for that scenario, so I can address this in the documentation and ensure it does not result in execution errors.

tniessen reviewed (commented):
I tend to agree with Ben that a more general mechanism would be preferable over adding one or even multiple options and then yet another API.

bnoordhuis commented:
I don't think this is the right direction. I hope we can agree the callback-based filter approach is superior when it comes to flexibility.

Adding a `depth` option that's subsequently subsumed by a better approach adds maintenance burden (can't really remove it once it's out there) and user confusion ("which one should I use?").

prettydiff commented:
@bnoordhuis late yesterday I removed the depth option and replaced it with a filter option, which is a function.  The filter function receives either a path or dirent object as the first argument an a number as the second argument representing the depth of directories from the base path.  All tests and documentation are caught up to expect an option named filter with conforming function.

LiviaMedeiros commented on lib/internal/fs/promises.js:
```diff
@@ -790,50 +782,56 @@ async function readdirRecursive(originalPath, options) {
-            direntPath,
-            await binding.readdir(
+        const direntPath = pathModule.join(path, dirent.name);
+        if (typeof options.filter !== 'function' || options.filter(direntPath, depth)) {
```
```suggestion
        if (typeof options.filter !== 'function' || options.filter(dirent, depth)) {
```
IMHO if `option.withFileTypes` is truthy, it would make sense to pass `dirent` directly, wdyt?

LiviaMedeiros commented on lib/internal/fs/promises.js:
```diff
@@ -790,50 +782,56 @@ async function readdirRecursive(originalPath, options) {
-              kUsePromises,
-            ),
-          ]);
+        if (typeof options.filter !== 'function' || options.filter(direntPath, depth)) {
```
```suggestion
        if (typeof options.filter === 'function' && !options.filter(direntPath, depth))
          continue;
```
If I'm reading this part correctly, we can invert this `if` to reduce nesting.

LiviaMedeiros reviewed (commented):
The newer implementation of filtering function looks cleaner, I like it!

Also, maybe we can consider overloading the `recursive` option instead of introducing the filter as separate one? It might looks like this:
```mjs
readdir(mydir); // not recursive
readdir(mydir, { recursive: false }); // not recursive
readdir(mydir, { recursive: true }); // unconditionally recursive
readdir(mydir, { recursive: (path, depth) => depth < 3 }); // recursive with filter

// instead of
readdir(mydir, { recursive: true, filter: (path, depth) => depth < 3 }); // recursive with filter
readdir(mydir, { filter: (path, depth) => depth < 3 }); // not recursive
```

This is not how it works for [`fs.cp`](https://nodejs.org/api/fs.html#fspromisescpsrc-dest-options), but it doesn't look like a breaking change and won't introduce an option that is useful only when another boolean option is enabled.

LiviaMedeiros reviewed (commented):
With a few fixes, documentation and `fs/promises` implementation look good to me.

The depth checking in `lib/fs.js`, in my opinion, should be rewritten using same logic as in promises, with incrementing `depth` as number instead of doing math around `pathModule.sep`.
I guess it can be done by changing `pathQueue` into array of `[pathOrDirent, depth]` pairs? Roughly like this:
```diff
-const pathsQueue = [basePath];
+const pathsQueue = [basePath, 1];

-function read(path) {
+function read({ 0: path, 1: depth }) {

-ArrayPrototypePush(pathsQueue, resultPath);
+ArrayPrototypePush(pathsQueue, [resultPath, depth + 1]);
```

Also, it would be nice to have an example in documentation that demonstrates this feature. For example, ignoring directories with names starting with `.` and limiting `depth` to 3.
Of course, this can be added in separate PR.

<hr />

Unless there are any objections, marking it as https://github.com/nodejs/node/labels/semver-minor and https://github.com/nodejs/node/labels/notable-change, so it can be described in release notes.

prettydiff commented:
This looks like its going to build... the larger linux build is right at 3 hours of effort.  I request guidance about how to proceed with the unrelated removed code from promises.js and then I believe this will be ready.  Here is the commit where I first noticed it: https://github.com/nodejs/node/pull/49255/commits/4ee991561232289dca281733f48b6e871846eb73

LiviaMedeiros commented:
For the future pull requests I'd highly recommend to always start by [creating new branch](https://github.com/nodejs/node/blob/main/doc/contributing/pull-requests.md#step-2-branch).

For this one, you can squash commits into one and change first commit message.

1. Restore the deleted part of code. On upstream it was deleted in https://github.com/nodejs/node/commit/c5cad084760366886ed014a80fa991c79f2f8be0, so it can be found earlier: https://github.com/nodejs/node/blob/33710e7e7d39d19449a75911537d630349110a0c/lib/internal/fs/promises.js#L285-L292

2. Soft-reset to the parent commit (that could be base for separate branch):
```console
$ git reset --soft 7fa9316973403c6f5ea8e3f5b97e7c6ff13e5c64~1
```
3. Commit with newer message that reflects how this feature works now:
```
$ git commit -a -m "fs: add filter callback to recursive fs.readdir"
```
4. Force-push:
```
$ git push --force-with-lease
```
For more flexible way to do 2 and 3 steps, I'd recommend [`git rebase`](https://git-scm.com/docs/git-rebase) in interactive mode.

Don't worry about commits that were added to upstream after 41a3a1daa28ba1431fe3e7d2c0d15f6e9a816b95; there is no need to merge them as long as there are no really conflicting changes.

aduh95 commented on doc/api/fs.md:
```diff
@@ -1304,7 +1304,11 @@ changes:
   * `withFileTypes` {boolean} **Default:** `false`
-  * `recursive` {boolean} **Default:** `false`
+  * `recursive` {boolean|Function} **Default:** `false`
+    * `name` {string} A given directory file system artifact name.
```
```suggestion
    * `name` {string|fs.Dirent} A given directory file system artifact name or the `Dirent` instance associated with it.
```

LiviaMedeiros replied:
Or also it can be `Buffer`, depending on `encoding` option
```suggestion
    * `name` {string|Buffer|fs.Dirent} A given directory file system artifact name or the `Dirent` instance associated with it.
```

LiviaMedeiros reviewed (commented):
The code LGTM.

The documentation needs additions to YAML changelogs, like this one: https://github.com/nodejs/node/blob/c42f47c56c164f523c18ed355016b4cab9259fd4/doc/api/fs.md?plain=1#L1237-L1244

The `version` field for new entries should have `REPLACEME` as value.

In the examples, I think it would be useful to show results of calling these functions, mainly to help readers to understand how exactly directories are counted (and what exactly happens when they are filtered out).
Roughly, as user I'd like to see what `await readdir('a/b', (name, depth) => name !== 'c' && depth < 3);` would do if we have `a/b/c/d/e` and `a/b/0/1/2/3/4/5` directories.

It would demonstrate otherwise unobvious:
- that `name` of directory is its own name, not relative path to original directory nor absolute path
- from where `depth` starts and where it ends
- is filtered out directory itself included in results or not

The code starts counting depth from 1, is there a reason for that or we can start from 0?

prettydiff commented:
> Roughly, as user I'd like to see what await readdir('a/b', (name, depth) => name !== 'c' && depth < 3); would do if we have a/b/c/d/e and a/b/0/1/2/3/4/5 directories.

In that case I believe there should be a list of ['a/b/c', 'a/b/0', 'a/b/0/1', 'a/b/0/1/2'].  The logic only limits what to traverse and not what to populate, and I add some explicitness about this to the documentation.  I also modified the depth counter to start at 0 instead of 1.  So anything that is an immediate child of the base path is depth 0.  I also added code comments to the documentation explaining exactly what is limited in the given example code samples.

LiviaMedeiros commented:
Things that should be addressed before approval:

- https://github.com/nodejs/node/pull/49255#discussion_r1313826884
`name` type in documentation
- https://github.com/nodejs/node/pull/49255#discussion_r1301658685
In Promises API version, it should be possible to provide async function.
Code-wise it means:
  - in Promises version we can simply `await` it
  - in Callback and Sync versions we should throw if returned value is `Promise`
An example of what-should-we-throw-if-promise can be found in https://github.com/nodejs/node/blob/5b31ff112783f97d12763a6516fcdadb7fa046ec/lib/internal/fs/cp/cp-sync.js#L67-L69
  - the documentation of Promises version should mention that it can return `Promise` that resolves with truthy/falsy value

> In that case I believe there should be a list of ['a/b/c', 'a/b/0', 'a/b/0/1', 'a/b/0/1/2'].

I still think, having list like this in code example in documentation will be helpful. :)
But this is non-blocking concern; and we can address code example in follow-up PR, if needed.

LiviaMedeiros commented:
@prettydiff There is a merge conflict that prevents CI from starting, would you mind to squash and force-push it again, please?

The resulting commit should lie on top of 41a3a1daa28ba1431fe3e7d2c0d15f6e9a816b95 (right now the commit history has the 7fa9316973403c6f5ea8e3f5b97e7c6ff13e5c64 from initial version that's supposed to be squashed with subsequent commits).

prettydiff committed:
fs: add new option depth to readdir which regulates depth of recursion

prettydiff committed:
Merge branch 'main' into main

LiviaMedeiros commented on test/sequential/test-fs-readdir-recursive.js:
```diff
@@ -79,99 +79,349 @@ function createFiles(path, fileStructure) {
 
+const results = [
+  [
+    'a', 'a/bar', 'a/foo', 'aa', 'aa/bar', 'aa/foo',
```
```suggestion
    'a', 'a/a', 'a/bar', 'a/foo', 'aa', 'aa/bar', 'aa/foo',
```
There are test failures, we should expect `'a/a'` directory in outputs where it's not filtered out.

Jarred-Sumner commented:
This API will be very difficult to make performant. It means every directory returned must be converted into a string. It would work better to have a deny list, an allow list, a glob to match against, or something which can return a boolean without a  function.
